lib/internal/util/comparisons.js

```diff
@@ -223,12 +257,22 @@ function innerDeepEqual(val1, val2, strict, memos) {
     }
     return keyCheck(val1, val2, strict, memos, kIsMap);
   } else if (isAnyArrayBuffer(val1)) {
-    if (!areEqualArrayBuffers(val1, val2)) {
+    if (!isAnyArrayBuffer(val2) || !areEqualArrayBuffers(val1, val2)) {
       return false;
     }
-  }
-  if ((isBoxedPrimitive(val1) || isBoxedPrimitive(val2)) &&
-    !isEqualBoxedPrimitive(val1, val2)) {
+  } else if (isBoxedPrimitive(val1)) {
+    if (!isEqualBoxedPrimitive(val1, val2)) {
+      return false;
+    }
+  } else if (ArrayIsArray(val2) ||
```

Comment:
why check isArray of val2 way down here, instead of only up by the place where val1 is checked (as in #30743)?

Reply:
It's less CPU work.

In #30743 all entries have to check both sides of both types. But if the arguments are e.g., sets, we'd only have to know that one side is not a set and then continue until we match sets. There we verify the other side.
If it's neither of the explicitly listed entries, we'll verify that the right side is non of the already tested types.